jumpscale/sals/nettools/__init__.py

```diff
@@ -375,98 +459,240 @@ def get_host_by_name(dnsHostname: str):
+            specified a absolute localpath (start with /) append_to_home will have no effect. Defaults to False.
+        name_from_url (bool, optional): if set to true, localpath will treated as a dir, and will try to get
+            the file name from the url or fallback to auto generated name. Defaults to True.
+
+    Raises:
+        PermissionError: [description]
+        FileNotFoundError: [description]
+        FileExistsError: [description]
+        ValueError: [description]
+        URLError: [description]
+
+    Returns:
+        namedtuple: namedtuple('DownloadResult', ['localpath', 'content', content_length])
+            - localpath (pathlib.Path)
+            - content (bytes): only if localpath is None else it will be None always
+            - content_length: the content length as returned from the response headers
+
+    Todo:
+        - better performance with Multi-Threaded
+        - support resume
+
+    Examples:
+        # use default values for args will download the url to cwd and get the name from the url,
+        # if the file already exists, it will overwrited.
+        >>> nettools.download('https://www.7-zip.org/a/7z1900-extra.7z')
+        DownloadResult(localpath=PosixPath('/home/sameh/projects/js-ng/7z1900-extra.7z'), content=None, content_length='929117')
+
     """
-    if not url:
-        raise Value("URL can not be None or empty string")
-    if not localpath:
-        raise Value("Local path to download the url to can not be None or empty string")
-    filename = ""
-    if localpath == "-":
-        filename = "-"
-    if jumpscale.sals.fs.exists(localpath) and jumpscale.sals.fs.is_dir(localpath):
-        filename = jumpscale.sals.fs.join_paths(localpath, jumpscale.sals.fs.basename(url))
-    else:
-        if jumpscale.sals.fs.is_dir(jumpscale.sals.fs.dirname(localpath)):
-            filename = localpath
-        else:
-            raise Value("Local path is an invalid path")
-
-    if not jumpscale.sals.fs.exists(filename):
-        overwrite = True
-
-    if overwrite:
-        if username and passwd and jumpscale.tools.http.urllib3.util.parse_url(url).scheme == "ftp":
-            url = url.split("://")[0] + "://%s:%s@" % (username, passwd) + url.split("://")[1]
-        response = jumpscale.tools.http.get(
-            url, stream=True, auth=jumpscale.tools.http.auth.HTTPBasicAuth(username, passwd)
-        )
-        response.raise_for_status()
-        if filename != "-":
-            with open(filename, "wb") as fw:
-                for chunk in response.iter_content(chunk_size=2048):
-                    if chunk:  # filter out keep-alive new chunks
-                        fw.write(chunk)
-            return
+    DownloadResult = namedtuple("DownloadResult", ["localpath", "content", "content_length"])
+    file_name = ""
+
+    try:
+        parsed_url = urlparse(url)
+    except ValueError:
+        raise
```

Comment:
we can log the exception here and use js exceptions too

Reply:
any specific way I should go with it? specific log file I should write to, or specific framework to use. referring me to an example of the proper style or method used to log the exceptions in the project also would be great.